prov/sockets/src/sock_progress.c

```diff
@@ -228,9 +235,19 @@ static struct sock_pe_entry *sock_pe_acquire_entry(struct sock_pe *pe)
 	return pe_entry;
 }
 
+static int sock_pe_is_completion_reported(struct sock_pe_entry *pe_entry)
+{
+	int completion_reported = ofi_atomic_inc32(&pe_entry->completion_reported);
+	return (completion_reported != 1);
+}
+
 static void sock_pe_report_send_cq_completion(struct sock_pe_entry *pe_entry)
 {
 	int ret = 0;
+
+	if (sock_pe_is_completion_reported(pe_entry))
+		return;
```

Comment:
This only protects against inserting multiple completions into the CQ. This check (and the one to report_recv_cq_completion) need to move up to the calling functions to ensure that the entire completion is only reported once.  Otherwise, we can bump the completion counter multiple times in cases where a counter is used in place of a CQ.

This means moving the check for ```if (pe_entry->completion_reported)``` into: sock_pe_report_send_completion(), sock_pe_report_recv_completion(), sock_pe_report_write_completion(), sock_pe_report_read_completion(), sock_pe_report_remote_read(), and wrapping around the call pair sock_pe_report_remote_write() + sock_pe_report_mr_completion().

I think that covers this, but we'll need to ensure that we get this right.